### PR TITLE
Replace ShaderBindings with new ResourceLayout structure for Vulkan

### PR DESCRIPTION
**Pull request by gdkchan**
Since Vulkan requires the creation of `PipelineLayout`s that indicates the bindings the shader uses, and for the creation of descriptor sets, we currently have a `ShaderBindings` structure where the bindings number are passed. Each shader stage has one of such structs.

This change removes the `ShaderBindings` structure and replaces it with something more flexible, a `ResourceLayout` structure that has not only the binding numbers, but also the type of resource at the binding (buffer, texture, image etc), the access type (read-only or read/write), the shader stages where it is used at, and the count for array descriptors. The `ResourceLayout` is used to create the `PipelineLayout` on Vulkan, rather than it having 2 hardcoded layouts (one that was used for game shaders, and another that was used for helper shaders from the backend called "minimal layout").

The main goal of this change is allowing transform feedback and geometry shader emulation to be implemented in a "clean" way. Since we need to reserve additional storage buffers  for TFB and GS emulation, the `PipelineLayout` also needs to be different. This change allows this to be done in a simple way.

`DescriptorSetUpdater.UpdateAndBind` was also greatly simplified. Instead of 3 loops now it only needs one. The method used to check if a texture is a buffer texture is also simpler and more robust. This can give a minor performance improvement.

Testing is welcome.

### CODE REVIEW

**Review by riperiperi — APPROVED** [review]
Nice, the bindings for the HelperShaders are definitely a lot cleaner. Removing the stages loop from the descriptor set updater is a nice bonus. Everything appears to be in working order when running a bunch of games.

File: src/Ryujinx.Graphics.GAL/ResourceLayout.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        public ReadOnlyCollection<ResourceDescriptor> Descriptors { get; }
+
+        public ResourceDescriptorCollection(ReadOnlyCollection<ResourceDescriptor> descriptors)
+        {
+            Descriptors = descriptors;
+        }
+
+        public override int GetHashCode()
+        {
+            HashCode hasher = new HashCode();
```
[the remark this answers is not in the document]
Comment (gdkchan) [reply src/Ryujinx.Graphics.GAL/ResourceLayout.cs:111]: Right now it should only be called once (when the shader program object is created). Unless 'ConcurrentDictionary.GetOrAdd' calls it more than once internally.

**Comment by MelonSpeedruns** [comment]
What games should we test, and what should we test for?

**Comment by gdkchan** [comment]
> What games should we test, and what should we test for?

This is not the type of change that only affects specific games. If something breaks it should break pretty much any game. So you can test pretty much anything, but I don't think it needs extensive testing. Also I optimized the descriptor set update method a bit, but I don't know if it will result in noticeable performance improvement, I guess it depends on the game and how often and how many descriptors it needs to update.

**Comment by MelonSpeedruns** [comment]
> > What games should we test, and what should we test for?
> 
> This is not the type of change that only affects specific games. If something breaks it should break pretty much any game. So you can test pretty much anything, but I don't think it needs extensive testing. Also I optimized the descriptor set update method a bit, but I don't know if it will result in noticeable performance improvement, I guess it depends on the game and how often and how many descriptors it needs to update.

I will try a few games to see any performance improvement then, thank you!

**Review by marysaka — APPROVED** [review]

**Pull request merged**